tests/e2e/stablehlo_ops/CMakeLists.txt

```diff
@@ -471,6 +474,168 @@ iree_check_single_backend_test_suite(
 
 ### BAZEL_TO_CMAKE_PRESERVES_ALL_CONTENT_BELOW_THIS_LINE ###
 
+# TODO(scotttodd): could add `CONDITION` options to
+#   iree_check_single_backend_test_suite that would allow for
+#   tests being defined but marked DISABLED, rather than wrap these with
+#   all-or-nothing CMake `if(...)` code...?
+if(IREE_ENABLE_ASAN)
+iree_check_single_backend_test_suite(
+  NAME
+    check_llvm-cpu_local-task_asan
```

Comment:
Attaching a comment to this so it doesn't get lost in the PR diff.

Reply:
Thanks for taking this on. I really like where this is going!

My only hesitation is how this moves the decision-making of "what to run with which sanitizers" to the test directories. I am afraid that the decision of which sanitizers to apply where, is not best taken at this granularity, in these locations. The one thing that the old way had for itself, was keeping the topic of sanitizers mostly orthogonal to the test source directories.

This concern would not be fully addressed by what is proposed in the comment above.

Possible ways out:
1. If we want to keep 1:1 test coverage with the existing way that we are running ASan tests with and without ASan-in-bytecode-modules, then maybe we can plumb sanitizer options through CMake/Bazel functions like `iree_check_single_backend_test_suite` and have them ultimately cause the separate `_asan`-suffixed tests, and with plumbing of the option all the way down to `iree_bytecode_module`, we can have that take care of the additional compiler flags, like in the existing code iirc. It is nice to keep these additional compiler flags a detail within `iree_bytecode_module` rather than leak the detail all the way up the build system.
2. If we don't want to keep 1:1 parity of test coverage, then we don't need plumbing and have these CMake functions just do that based on IREE_ENABLE_ASAN. What we lose then is ASan coverage for code paths specific to embedded-library modules. Maybe that's OK: that's a small amount of low-churn code that's shared among all workloads so at this point it's just unlikely that ASan coverage ever catches a bug in that specifically.

WDYT @benvanik

Reply:
Note: whichever way this goes, I will very much enjoy only having to flip `IREE_ENABLE_ASAN` when I want to run tests with ASan. And I don't remember ever needing specifically ASan only in C/C++ code and not in bytecode modules.

Reply:
The verbosity aspect is one aspect that I agree we can compromise on, which is why I didn't make it the primary angle in my previous comment. Instead of verbosity, I focused on "where is the decision point" to have a ASan and/or a TSan variant for some test. My main concern is that if, as the current form of this PR does, it's moving to the test definition in each test directory's CMakeLists, then we are by default going to have inconsistent and arbitrary choices of what tests have TSan/ASan. Moving the decision to helpers is thus not (only) avoiding verbosity, it is (primarily) about avoiding scattering the decision points.

Reply:
right. sounds like we now have the contours of the discussion in place .... do we need a video call next? @benvanik do you want to be involved ?

Reply:
Had another idea - we could add a `iree_check_cpu_test_suite` that looks at the `IREE_ENABLE_*SAN` options and decides which tests to create. Could also have `iree_check_cuda_test_suite` that creates stream/graph/etc. variants. The main thing I want to avoid is each target wanting to add its own special handling to the base function.

Reply:
(I'm not going to support Bazel for such new test suite helper functions though -- if someone cares about running test suites on Bazel they should help maintain them)

Reply:
Yes, IIUC this is what I also had in mind above in point 2. under "possible ways out".

Reply:
> (I'm not going to support Bazel for such new test suite helper functions though -- if someone cares about running test suites on Bazel they should help maintain them)

I don't expect that we need to care about Bazel+ASAN. And if we did, this direction would actually work well for it: by making ASAN a detail of rules like `iree_check_cpu_test_suite`, we would be able to do the right separately in the Bazel vs CMake implementations. So, where the CMake implementation would read the value of `IREE_ENABLE_ASAN`, the Bazel implementation would do the equivalent Bazel things based on config_settings etc.  (Give or take the gotchas about config_settins and Bazel macros but you get the idea).

Reply:
Yeah, these test suites _can_ still work with Bazel, but I have a few similar refactorings in mind that compound the effort needed.

For example, I also want to rework the inlined SRCS lists / globs to something more like:
```
ALL_TESTS = [ ... ]
iree_check_test_suite(
  name = "cpu"
  tests = ALL_TESTS
  xfail_tests = [ ... ]
  exclude_tests = [ ... ]
)
iree_check_test_suite(
  name = "gpu_vulkan"
  tests = ALL_TESTS
  xfail_tests = [ ... ]
  exclude_tests = [ ... ]
)
```
That can be done with filegroups, globs, and the converter script... but these tests aren't providing enough value for all that design and maintenance work IMO.

Reply:
Pushed a commit with a sketch for a helper function. Seems awkward. Argument forwarding in some way could help (https://gitlab.kitware.com/cmake/cmake/-/issues/22157).

I think a better simplification would be reducing the 'SRCS' copy/paste and then keeping the suites explicit in the files that want them.

Reply:
I'm going to be out from tomorrow through Monday. If you want to discuss this, we can do it today or next Tuesday. Without going into implementation details, the high-level trait that I care about is that ASan/TSan flavors of tests are created automatically, not as the result of the owner of some directory explicitly thinking about it.

Reply:
> the high-level trait that I care about is that ASan/TSan flavors of tests are created automatically, not as the result of the owner of some directory explicitly thinking about it.

I disagree there. Implicit logic in build system helper functions have been a continuing source of misconfigured CI builds and infrastructure maintenance overhead. Explicit is better than implicit. I'd like to remove the entire check test framework at this point (or at least dramatically rework how we use it), but I haven't finished putting a replacement together yet.

Reply:
Let's have a video chat when you have time!

Reply:
Discussed offline. I'll try moving the branching from `iree_check_cpu_test_suite` back into `iree_check_single_backend_test_suite`. As long as we have test suites tightly coupled to the build system that sort of carve-out may continue to make sense for sanitizer builds.

With how I've been steering tests towards package installs I think we can make further simplifications down the line. Some progress is better than being deadlocked.

Reply:
Made the changes we discussed. How does this look now?